### PR TITLE
[pytorch][mobile] custom build script

### PR DESCRIPTION
**Pull request by ljk53**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30144 [pytorch][mobile] custom build script**

Summary:
Create script to produce libtorch that only contains ops needed by specific
models. Developers can use this workflow to further optimize mobile build size.

Need keep a dummy stub for unused (stripped) ops because some JIT side
logic requires certain function schemas to be existed in the JIT op
registry.

Test Steps:
1. Build "dump_operator_names" binary and use it to dump root ops needed
by a specific model:
```
build/bin/dump_operator_names --model=mobilenetv2.pk --output=mobilenetv2.yaml
```

2. The MobileNetV2 model should use the following ops:
```
- aten::t
- aten::dropout
- aten::mean.dim
- aten::add.Tensor
- prim::ListConstruct
- aten::addmm
- aten::_convolution
- aten::batch_norm
- aten::hardtanh_
- aten::mm
```
NOTE that for some reason it outputs "aten::addmm" but actually uses "aten::mm".
You need fix it manually for now.

3. Run custom build script locally (use Android as an example):
```
SELECTED_OP_LIST=mobilenetv2.yaml scripts/build_pytorch_android.sh armeabi-v7a
```

4. Checkout demo app that uses locally built library instead of
downloading from jcenter repo:
```
git clone --single-branch --branch custom_build git@github.com:ljk53/android-demo-app.git
```

5. Copy locally built libraries to demo app folder:
```
find ${HOME}/src/pytorch/android -name '*.aar' -exec cp {} ${HOME}/src/android-demo-app/HelloWorldApp/app/libs/ \;
```

6. Build demo app with locally built libtorch:
```
cd ${HOME}/src/android-demo-app/HelloWorldApp
./gradlew clean && ./gradlew assembleDebug
```

7. Install and run the demo app.

In-APK arm-v7 libpytorch_jni.so build size reduced from 5.5M to 2.9M.

Differential Revision: [D18612127](https://our.internmc.facebook.com/intern/diff/D18612127)

### CODE REVIEW

**Comment by ljk53** [comment]
@iseeyuan do you know why MobileNetV2 contains "aten::addmm" operator in the instruction list but it ended up calling "aten::mm"? Do we need first look up JIT registry to "resolve" the operator binding before dumping?

**Comment by iseeyuan** [comment]
@ljk53 It may be caused by JIT optimization passes as well. Let me check if it's possible to dump the op list right before the interpreter running stage. It may be deeper in the graph executor.

**Comment by ljk53** [comment]
> @ljk53 It may be caused by JIT optimization passes as well. Let me check if it's possible to dump the op list right before the interpreter running stage. It may be deeper in the graph executor.

Is it possible to do different JIT optimization on mobile v.s. on server (might depend on input as well)? If we dump operator list on server how can we make sure it covers all possible optimization passes? cc: @zdevito

**Comment by ljk53** [comment]
BTW, I think this script can be reviewed independently of the op-dump issue. We can fix the issue separately.

**Comment by ezyang** [comment]
Can we put the docs in a place more durable than a PR description?

**Comment by iseeyuan** [comment]
> > @ljk53 It may be caused by JIT optimization passes as well. Let me check if it's possible to dump the op list right before the interpreter running stage. It may be deeper in the graph executor.
> 
> Is it possible to do different JIT optimization on mobile v.s. on server (might depend on input as well)? If we dump operator list on server how can we make sure it covers all possible optimization passes? cc: @zdevito

It may be different from the final bytecode in full JIT. For lite interpreter we dump the bytecode from the original module, without JIT opt passes. We have one set of bytecode independent of input-based optimizations. The performance should not be affected significantly (cc @zdevito ). The actual performance difference is pending to be measured in detail.

**Comment by iseeyuan** [comment]
Probably a stupid idea: it may not be hard to disable opt passes in mobile (in general, not just for lite interpreter). In that sense we don't need to worry about the input-dependent passes and different ops introduced by each pass. Not sure if it can be at least a short-term solution.

File: CMakeLists.txt [hunk 1]
```diff
@@ -205,6 +205,8 @@ cmake_dependent_option(
   "MSVC" OFF)
 
 set(ONNX_NAMESPACE "onnx_torch" CACHE STRING "A namespace for ONNX; needed to build with other frameworks that share ONNX.")
+set(SELECTED_OP_LIST "" CACHE STRING
+    "Path to the yaml file that contains the list of operators to include for custom build. Include all operators by default.")
```
Comment (ezyang) [inline CMakeLists.txt]: For doc discoverability, would be good to have a link to the relevant docs here

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**